From patchwork Tue Sep  6 19:48:46 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968071
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 7F795ECAAA1
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:49:24 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id 658B394000C; Tue,  6 Sep 2022 15:49:09 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id 5DE1694000A; Tue,  6 Sep 2022 15:49:09 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id EEAD794000E; Tue,  6 Sep 2022 15:49:08 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0010.hostedemail.com
 [216.40.44.10])
	by kanga.kvack.org (Postfix) with ESMTP id 9E2D994000B
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:49:08 -0400 (EDT)
Received: from smtpin13.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay06.hostedemail.com (Postfix) with ESMTP id 7C697AB262
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:08 +0000 (UTC)
X-FDA: 79882699176.13.339E6FD
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf17.hostedemail.com (Postfix) with ESMTP id ED7F240071
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:49:07 +0000 (UTC)
Received: from pps.filterd (m0246629.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286Id90Y024660;
	Tue, 6 Sep 2022 19:48:54 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=ZaC7lqQRQkSDIkJGa0g+/q1+WSqSUKrFo9yctwdc7aM=;
 b=kmoasXSH5EA+TyzdZ1hZzpKrbnSFsJPTlq+4CLOWdJXySoFrfipiQmwDak5n7KYP86oU
 j0pOMQadcR0rNi3i2xys2IjREJZCL1exEDYQLvZv8+UVmGzo3yfT6i7EGvB7FxDDcpYp
 O1ielOPJLqdBtmWD/U2GvNdDBTtBBydM2KeQ7alclJTzUvmNPGnYRckpAdFpJVoGtIdh
 69ulggCn2Bc+tL3e0Z1PlIwmS6AwcatiU0BfWIheV1I+uexF3NcA78J4C2rjGpZRjyf4
 t7cklqbtf36sMy3DzK1jeAPf33d4Vfq6rHVz6z1bQ3CzemtrssncyTzP5i8np3SaM/Qg ug==
Received: from iadpaimrmta01.imrmtpd1.prodappiadaev1.oraclevcn.com
 (iadpaimrmta01.appoci.oracle.com [130.35.100.223])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbxtaet55-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:48:54 +0000
Received: from pps.filterd
 (iadpaimrmta01.imrmtpd1.prodappiadaev1.oraclevcn.com [127.0.0.1])
	by iadpaimrmta01.imrmtpd1.prodappiadaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286H8KXJ023727;
	Tue, 6 Sep 2022 19:48:52 GMT
Received: from nam12-dm6-obe.outbound.protection.outlook.com
 (mail-dm6nam12lp2174.outbound.protection.outlook.com [104.47.59.174])
	by iadpaimrmta01.imrmtpd1.prodappiadaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc9kr9w-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:48:52 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=kQ0yGwd4lf9ptbcSBS688qoBJwHFTdo8V8Y73QG2XIf4GrRGyn9IwpcJZdAJbZ/YpDhjMu9DlVBipnWUwsYDdsQDmDw+VqQXQnS01/Z0FHmGEBI1GVouqjREIX+0PcLStYbhSRmHepJtCux9ZTzffgSyunvZZSPhMxyaZX9U0yus/DbjPP9HY4Y03HLwl3dq0NERlSvjQibJGl82SLJc7CiT8yOF4nXujV7WCbwOhjUJjaWT0i1Nw/RE5cVm1XdO9K2kMZRLx0ARvlOmKtfIIdm1fgLSeuMeb8LIqBMkY8rZNNtil4he/0p+Je8JIGkU8mOE6tYQD9h3bdoDHiv/4g==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=ZaC7lqQRQkSDIkJGa0g+/q1+WSqSUKrFo9yctwdc7aM=;
 b=LztuM1+AK/RmLLJSRxN7Q7U5YQeUR07HJT57M+t7b9uGL/OCyh8Pu9sjtH9RCHJ1HeYbMk5tDErUAfdI7HiloOno7JHsnhC2wi+nChmcS3I++1WaubQpFk9xEqLAKgAKr5iHHgVxu93ySQ8jHbHOJzIiczr7lUI3cud87yDgHO3VUrjC1KyRsCQD0jZrNPoGzImuBVZxmSS0ezxfSMyat83o/4pASwwaBaeCGYFEkZVkpni4TOdJwihYNgdZ2pJ5K1Cvcy5fD+CWZS8NA/8AilTDlziDKIhVeGsSWnagKB7/RSjlQ5tAEwkxrwWzaSMCxoQrZED6dmCL4SlaLs8OVQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=ZaC7lqQRQkSDIkJGa0g+/q1+WSqSUKrFo9yctwdc7aM=;
 b=HI+dWHjUVeAru2QpNwO20Jjlg5QjwG7tIFjjBEsxHkIunDELAoS+4wu4eo2b7l6kL10JXwoWXBgnOyU1VHZmOjXxJjTRnqLqSXSG47070tWoTR9wqvq/L2KgPbdqOW+jOXhxXy3i4FpX4GylCfRIpJcTE2f3Xr6PZpUdsXrpiwQ=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by DM4PR10MB6183.namprd10.prod.outlook.com (2603:10b6:8:8b::22) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.10; Tue, 6 Sep
 2022 19:48:50 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:48:50 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: "Matthew Wilcox (Oracle)" <willy@infradead.org>,
        Liam Howlett
	<liam.howlett@oracle.com>,
        Vlastimil Babka <vbabka@suse.cz>,
        David
 Hildenbrand <david@redhat.com>,
        Davidlohr Bueso <dave@stgolabs.net>
Subject: [PATCH v14 09/70] mm: add VMA iterator
Thread-Topic: [PATCH v14 09/70] mm: add VMA iterator
Thread-Index: AQHYwimt3pVVX5iKe0W6kP0G9E2e6g==
Date: Tue, 6 Sep 2022 19:48:46 +0000
Message-ID: <20220906194824.2110408-10-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: e496790d-5e0d-438c-e088-08da9040d243
x-ms-traffictypediagnostic: DM4PR10MB6183:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 sSh0D1WGAwAXcqxIh+np1d64WoL/1ZhhDec3KrOUwMzNGRVibZ6pEhoXXaDkbXUVkGW3Et+4/AkDgirNFny1Ie1CSdAF1Fx0EfkRmeY8xKFf5t+xAUqaaNnjHW0ui8s0fPq6q4ddA3KGenO5zaVQ0SdHoIsO5JB/bfeUxwXAyboHBu7Kb40zxKoE2g9d/A62FLuIRYkU4KcnWNnuVdCZYZm02LXGrE0bQ8D+066w+HxHlJ3CaWKvEa1RlwaHDWs6b20tDMXa1efXo653r7skd8mvjLK4tfos+HgsDTV0vTtbG5ukBCMeK7B/p4zBqqrVAtL0jSjpswZJGWWyWRbwWlv29FYAq50ILlBJWmlUv4R1s3pzrbTPinsUnY7hk0USXStgNxrNEiNuQZxqXlh8mU29ZvXQJW6SueJD9hPzJQayqfT1EYBwcB9iC9kUbgkgHlw7O+zq/+/DrhsuroM9yWfxK/7V5TLi9hAUkEghVZlzwzVVBcLGY5zxtmLii9EK/HNCqkq0nqs5nnBUJCbSz79D6mPBFGC+ezXf0+ef3bl6cuIjXyMQJul0HIWsIfrsfvZSMxvMXlP4J/KxIkifuj696f8loM/ugl806Ddjxe0dGXY8YiC8aD2fwom3c8jPdtDom8YTZykmJuusMwV5CHZTq2XvTQ9EUvr/+8c3iTbqrhjY+oAThQnX0wkv/cr+m+mHgAovEMZhbBSeexljo+VQALb+pk+Fu2zyIO92oMs1ssHkEEfZdSzCtFGjnaMEq6+ARa6i2TOJwHEKWLdNVg==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(39860400002)(366004)(376002)(396003)(346002)(136003)(66556008)(38100700002)(122000001)(8936002)(5660300002)(76116006)(66946007)(66446008)(66476007)(2906002)(83380400001)(8676002)(91956017)(36756003)(2616005)(186003)(86362001)(1076003)(110136005)(44832011)(71200400001)(6666004)(38070700005)(54906003)(26005)(64756008)(41300700001)(6506007)(6512007)(6486002)(316002)(478600001)(4326008);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?uV4dBUNhxWsl3sLSPXGs1mm?=
	=?iso-8859-1?q?XK2chBtmghFXUwrSK7Isfi7yxxBgYVx6Xjpl586VONqBNzowPGuBe5I1r2pf?=
	=?iso-8859-1?q?+DxSZWKJYJnBjU5b+kESysD2WWAn7KBcjzlciJR+7q0v1atSZzq5wLWy+fDv?=
	=?iso-8859-1?q?d+mIYWuu7xpFFV2HSs678DJ0DbKZZtMtgNm5gceAP5Qor4BPsKVGPuQv0IzG?=
	=?iso-8859-1?q?lEJvovBxI2jo8syup+wDsjYGKFlqWisoUCWp4G7xFH5KvDd605tzHtd+xr2W?=
	=?iso-8859-1?q?v9hlOPIFv598KdoxF23Aa5lw8g5B9hZUKlnP5cS6Ch6WVm62l66YqjYrWLVd?=
	=?iso-8859-1?q?7CijDqZlpsKDyWJJ+T4GCrkPHxK1XdFgq3RAyf96bCBlpYROY5hQpqdcOyJZ?=
	=?iso-8859-1?q?c1ImcfvzKZ41yXo7kmUYUmen6asDB/nZVGYJFE8653FZLMmMzR5qVGpuCb3S?=
	=?iso-8859-1?q?RXc4D/sJ1TPf/gH9RjujDkj+nqpEbcxnItQ2cTfksDdKwOP1CLnErzQvf6SC?=
	=?iso-8859-1?q?rclWuz5gpaLy3ZWPqPMxHSJWkooYY5ZV8Wp1sC/V1D5dR1vV7YHDSdkmgg/2?=
	=?iso-8859-1?q?4TKv+jOdYDqHtzlOzvOuziGnqRYLwYRJh1XE03MEQhzOhnf1lyWnqKsJ/JW+?=
	=?iso-8859-1?q?IRYZ9lTa2dKehVC7w4gtGyzUN+WA1FfEPwtrNPgRd0jge9PXPZ3cIeGURZu4?=
	=?iso-8859-1?q?/Ndr1iVpn1bvzx5Qi2A54FTxsLVKsBk8k+BpFDWCVh4sdvy4w+Xr6Wa0yzh+?=
	=?iso-8859-1?q?je4Jw065AbrYmYkTTvKi6VeKGT6VJc8wGB4uqXfFx12r7MrYX8d0EKjl7kAF?=
	=?iso-8859-1?q?n9IjycDGZMIudjtq3b0t8FBliq1ciEr6TdK1fV2OJHvdo2GhJsX+6BdQfKG/?=
	=?iso-8859-1?q?YMfbsXwovDwfKdO4SQyjZbYfTRR+WH1KcYGHquYSsxJYrRNuD66h0nxecEI0?=
	=?iso-8859-1?q?SqeZXmPElWHlnWKti04i2nuT5IosZw9F9yoRinZhzPj86CDR+nDp+v/gZraU?=
	=?iso-8859-1?q?T6/0fT/YX+yzmJxdmTS2Tmr4iLgGwrkB9WhA15EMacdQyXgGngWyR+nSdPk4?=
	=?iso-8859-1?q?qlGkvbnXkdPFsjHYW3CbRDJOTOxa8RVYMckHqmNj8Yez8zOguRV1dqySeM+/?=
	=?iso-8859-1?q?No9xZJ4VyL3UIRizBI8hGRdjvN3jIHsxTf5n1XsxmZhDtefMLuCpyCammQKB?=
	=?iso-8859-1?q?mzfFvZjPFPljNFbkc4Ozo0NwNr2BsGhFwOt6zUiotI1ypqj9egHhjrQ3y/KQ?=
	=?iso-8859-1?q?eQeEfOX4vvFtzLrEbGAHtv4C8m6WaZXhvyWZUyJDJrC/ucTt8RzDDA/MKOS9?=
	=?iso-8859-1?q?K9eeW4YVFh4Mhm7zbN/itSmb+Uo5EgsvkRIv+d0XQMKaDGQXFlZOF3nVLrCg?=
	=?iso-8859-1?q?x0RXcedlX7ZETpJvX9S4WhUiyJjry3RAktnTYFY1ji7vley6UCWN4niuUFZk?=
	=?iso-8859-1?q?7ImsNldLkojh3RJpc9lVvziJe0mg3fjxbnhLVY/N3/tM9+LGMOHYhFK5QeJw?=
	=?iso-8859-1?q?+U5TObMhEaFUg1PjkmZ0u7e8Gs/q8N2wXZNbgSbFFIkCdpX90ftR5ct530oS?=
	=?iso-8859-1?q?Vn1KvO6Takt33EQtkmXR7kQFRgIa4etfvSQvFhpOtmdWefic64qazkKpQpxe?=
	=?iso-8859-1?q?DHaX1MrrOGoFjC3LH32o5uJ67KLO7OzuxGHUV8g=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 e496790d-5e0d-438c-e088-08da9040d243
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:48:46.1319
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 Owx6BYMXnbZy3A8eqrAXVuuVaEGbdyPZ1MHFMlR5xIyTpITY/ggdLPFtqPUOE0P2GxWptskBFlIXzgC+Ow9lqA==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: DM4PR10MB6183
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 phishscore=0
 mlxscore=0 adultscore=0
 malwarescore=0 mlxlogscore=999 bulkscore=0 suspectscore=0 spamscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2207270000
 definitions=main-2209060091
X-Proofpoint-ORIG-GUID: PcS3GhWXAttdC2t4yJ7W6iMrKoUvfj7Q
X-Proofpoint-GUID: PcS3GhWXAttdC2t4yJ7W6iMrKoUvfj7Q
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493748;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=ZaC7lqQRQkSDIkJGa0g+/q1+WSqSUKrFo9yctwdc7aM=;
	b=lqXczWY8SutabhBQc6xFZoVZODfOOl0ZloGxkkR6M8qEedCGrMR/vAbpI6TY3I+cV3bEPM
	9eTZzwszQ3Ak1GL3gSde1ZaVl/qO/BxlTfoYL/ju6nKIz+cMehfCMNQTRjKfdZLIJ/3ezO
	0GfHMEpKvtJSNx2+vkAhiOwxdL7rk60=
ARC-Authentication-Results: i=2;
	imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=kmoasXSH;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=HI+dWHjU;
	spf=pass (imf17.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1");
	dmarc=pass (policy=none) header.from=oracle.com
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493748; a=rsa-sha256;
	cv=pass;
	b=hlEKwiH5ltCfjrfuIB9BTt26LhQFNUQrRSkonPqb5rwDcuKnC7yyvI20LD0bggNpqnFB+z
	aI+VuLNGJh43NKyhtsmIjORtaObxm9j6g0/nzPOV28sYYpEceZU5WOtRhdLlmsquPEtsgK
	W1OsbxI0QNcqKyExzOio/f2byaJRYB0=
X-Stat-Signature: mtj8tkrxhr3779bs8e1pbjtsucag6xrs
X-Rspam-User: 
X-Rspamd-Queue-Id: ED7F240071
Authentication-Results: imf17.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=kmoasXSH;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=HI+dWHjU;
	spf=pass (imf17.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1");
	dmarc=pass (policy=none) header.from=oracle.com
X-Rspamd-Server: rspam01
X-HE-Tag: 1662493747-871344
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Matthew Wilcox (Oracle)" <willy@infradead.org>

This thin layer of abstraction over the maple tree state is for iterating
over VMAs.  You can go forwards, go backwards or ask where the iterator
is.  Rename the existing vma_next() to __vma_next() -- it will be removed
by the end of this series.

Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz>
Reviewed-by: David Hildenbrand <david@redhat.com>
Reviewed-by: Davidlohr Bueso <dave@stgolabs.net>
---
 include/linux/mm.h       | 32 ++++++++++++++++++++++++++++++++
 include/linux/mm_types.h | 21 +++++++++++++++++++++
 mm/mmap.c                | 10 +++++-----
 3 files changed, 58 insertions(+), 5 deletions(-)

diff --git a/include/linux/mm.h b/include/linux/mm.h
index 5ee6e6d0c1c3..a458e5f73fee 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -661,6 +661,38 @@ static inline bool vma_is_accessible(struct vm_area_struct *vma)
 	return vma->vm_flags & VM_ACCESS_FLAGS;
 }
 
+static inline
+struct vm_area_struct *vma_find(struct vma_iterator *vmi, unsigned long max)
+{
+	return mas_find(&vmi->mas, max);
+}
+
+static inline struct vm_area_struct *vma_next(struct vma_iterator *vmi)
+{
+	/*
+	 * Uses vma_find() to get the first VMA when the iterator starts.
+	 * Calling mas_next() could skip the first entry.
+	 */
+	return vma_find(vmi, ULONG_MAX);
+}
+
+static inline struct vm_area_struct *vma_prev(struct vma_iterator *vmi)
+{
+	return mas_prev(&vmi->mas, 0);
+}
+
+static inline unsigned long vma_iter_addr(struct vma_iterator *vmi)
+{
+	return vmi->mas.index;
+}
+
+#define for_each_vma(__vmi, __vma)					\
+	while (((__vma) = vma_next(&(__vmi))) != NULL)
+
+/* The MM code likes to work with exclusive end addresses */
+#define for_each_vma_range(__vmi, __vma, __end)				\
+	while (((__vma) = vma_find(&(__vmi), (__end) - 1)) != NULL)
+
 #ifdef CONFIG_SHMEM
 /*
  * The vma_is_shmem is not inline because it is used only by slow
diff --git a/include/linux/mm_types.h b/include/linux/mm_types.h
index f5d6869912be..a249a7d5f5da 100644
--- a/include/linux/mm_types.h
+++ b/include/linux/mm_types.h
@@ -778,6 +778,27 @@ static inline void lru_gen_use_mm(struct mm_struct *mm)
 
 #endif /* CONFIG_LRU_GEN */
 
+struct vma_iterator {
+	struct ma_state mas;
+};
+
+#define VMA_ITERATOR(name, __mm, __addr)				\
+	struct vma_iterator name = {					\
+		.mas = {						\
+			.tree = &(__mm)->mm_mt,				\
+			.index = __addr,				\
+			.node = MAS_START,				\
+		},							\
+	}
+
+static inline void vma_iter_init(struct vma_iterator *vmi,
+		struct mm_struct *mm, unsigned long addr)
+{
+	vmi->mas.tree = &mm->mm_mt;
+	vmi->mas.index = addr;
+	vmi->mas.node = MAS_START;
+}
+
 struct mmu_gather;
 extern void tlb_gather_mmu(struct mmu_gather *tlb, struct mm_struct *mm);
 extern void tlb_gather_mmu_fullmm(struct mmu_gather *tlb, struct mm_struct *mm);
diff --git a/mm/mmap.c b/mm/mmap.c
index 5115eea6a0e6..20718645d82f 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -586,7 +586,7 @@ static int find_vma_links(struct mm_struct *mm, unsigned long addr,
 }
 
 /*
- * vma_next() - Get the next VMA.
+ * __vma_next() - Get the next VMA.
  * @mm: The mm_struct.
  * @vma: The current vma.
  *
@@ -594,7 +594,7 @@ static int find_vma_links(struct mm_struct *mm, unsigned long addr,
  *
  * Returns: The next VMA after @vma.
  */
-static inline struct vm_area_struct *vma_next(struct mm_struct *mm,
+static inline struct vm_area_struct *__vma_next(struct mm_struct *mm,
 					 struct vm_area_struct *vma)
 {
 	if (!vma)
@@ -1291,7 +1291,7 @@ struct vm_area_struct *vma_merge(struct mm_struct *mm,
 	if (vm_flags & VM_SPECIAL)
 		return NULL;
 
-	next = vma_next(mm, prev);
+	next = __vma_next(mm, prev);
 	area = next;
 	if (area && area->vm_end == end)		/* cases 6, 7, 8 */
 		next = next->vm_next;
@@ -2843,7 +2843,7 @@ static void unmap_region(struct mm_struct *mm,
 		struct vm_area_struct *vma, struct vm_area_struct *prev,
 		unsigned long start, unsigned long end)
 {
-	struct vm_area_struct *next = vma_next(mm, prev);
+	struct vm_area_struct *next = __vma_next(mm, prev);
 	struct mmu_gather tlb;
 
 	lru_add_drain();
@@ -3051,7 +3051,7 @@ int __do_munmap(struct mm_struct *mm, unsigned long start, size_t len,
 		if (error)
 			goto split_failed;
 	}
-	vma = vma_next(mm, prev);
+	vma = __vma_next(mm, prev);
 
 	if (unlikely(uf)) {
 		/*

